src/Messages/InvalidKeyMessages/TSInvalidIdentityKeyReceivingErrorMessage.m

```diff
@@ -78,20 +79,25 @@ - (void)acceptNewIdentityKey
         return;
     }
 
-    [[TSStorageManager sharedManager] saveRemoteIdentity:newKey recipientId:self.envelope.source];
-
-    // Decrypt this and any old messages for the newly accepted key
-    NSArray<TSInvalidIdentityKeyReceivingErrorMessage *> *messagesToDecrypt =
-        [self.thread receivedMessagesForInvalidKey:newKey];
-
-    for (TSInvalidIdentityKeyReceivingErrorMessage *errorMessage in messagesToDecrypt) {
-        [[TSMessagesManager sharedManager] handleReceivedEnvelope:errorMessage.envelope];
-
-        // Here we remove the existing error message because handleReceivedEnvelope will either
-        //  1.) succeed and create a new successful message in the thread or...
-        //  2.) fail and create a new identical error message in the thread.
-        [errorMessage remove];
-    }
+    // Saving a new identity mutates the session store so it must happen on the sessionStoreQueue
+    dispatch_async([OWSDispatch sessionStoreQueue], ^{
+        [[TSStorageManager sharedManager] saveRemoteIdentity:newKey recipientId:self.envelope.source];
+
+        dispatch_async(dispatch_get_main_queue(), ^{
+            // Decrypt this and any old messages for the newly accepted key
+            NSArray<TSInvalidIdentityKeyReceivingErrorMessage *> *messagesToDecrypt =
+                [self.thread receivedMessagesForInvalidKey:newKey];
+
+            for (TSInvalidIdentityKeyReceivingErrorMessage *errorMessage in messagesToDecrypt) {
+                [[TSMessagesManager sharedManager] handleReceivedEnvelope:errorMessage.envelope];
+
+                // Here we remove the existing error message because handleReceivedEnvelope will either
+                //  1.) succeed and create a new successful message in the thread or...
+                //  2.) fail and create a new identical error message in the thread.
+                [errorMessage remove];
+            }
+        });
+    });
```

Comment:
It looks like `acceptNewIdentityKey` is only being called from `MessagesViewController`.  Immediately after, `MessagesViewController` tries to resend a message.  Now that `acceptNewIdentityKey`, should it take a completion block so that we can ensure that the resend occurs after the acceptance is complete?